tests/custom_lookups/tests.py

```diff
@@ -647,3 +671,84 @@ def test_subquery_usage(self):
+        with register_lookup(Company.place, CustomStartsWith):
+            self.assertEqual(Company.name.field.get_lookup("sw"), None)
+            self.assertEqual(Company.place.field.get_lookup("sw"), CustomStartsWith)
+
+    def test_get_lookup_from_proxy_method(self):
+        with register_lookup(models.CharField, CustomStartsWith):
+            self.assertEqual(Company.name.get_lookup("sw"), CustomStartsWith)
+            self.assertEqual(Company.place.get_lookup("sw"), CustomStartsWith)
+        with register_lookup(Company.place, CustomStartsWith):
+            self.assertEqual(Company.name.get_lookup("sw"), None)
+            self.assertEqual(Company.place.get_lookup("sw"), CustomStartsWith)
+
+    def test_get_transform_from_proxy_method(self):
+        with register_lookup(models.CharField, Div3Transform):
+            self.assertEqual(Company.name.get_transform("div3"), Div3Transform)
+            self.assertEqual(Company.place.get_transform("div3"), Div3Transform)
+        with register_lookup(Company.place, Div3Transform):
+            self.assertEqual(Company.name.get_transform("div3"), None)
+            self.assertEqual(Company.place.get_transform("div3"), Div3Transform)
+
+    def test_class_lookup_with_filter(self):
+        with register_lookup(models.CharField, CustomStartsWith):
+            self.assertCountEqual(
+                Company.objects.filter(name__sw="a"), [self.obj1, self.obj2, self.obj4]
+            )
+            self.assertCountEqual(Company.objects.filter(name__sw="c"), [self.obj3])
+
+    def test_instance_lookup_with_filter(self):
+        with register_lookup(Company.name, StartsWith4Char):
+            self.assertCountEqual(
+                Company.objects.filter(name__swf="a"), [self.obj1, self.obj2]
+            )
+            self.assertCountEqual(Company.objects.filter(name__swf="c"), [self.obj3])
+            with self.assertRaises(FieldError):
+                Company.objects.filter(place__swf="c")
+
+    def test_instance_overide_class_lookups_with_filter(self):
+        with register_lookup(models.CharField, CustomStartsWith, lookup_name="start"):
+            with register_lookup(Company.place, StartsWith4Char, lookup_name="start"):
+                self.assertCountEqual(
+                    Company.objects.filter(place__start="f"),
+                    [self.obj1, self.obj3],
+                )
+                self.assertCountEqual(
+                    Company.objects.filter(name__start="a"),
+                    [self.obj1, self.obj2, self.obj4],
+                )
+            self.assertCountEqual(
+                Company.objects.filter(place__start="f"),
+                [self.obj1, self.obj3, self.obj4],
+            )
+
+    def test_instance_overide_instance_lookup(self):
+        with register_lookup(Company.name, CustomStartsWith, lookup_name="start"):
+            Company.name.register_lookup(StartsWith4Char, lookup_name="start")
```

Comment:
Use `try ... finally` and unregister this lookup to keep tests isolated.

Reply:
The registered lookups share the same `lookup_name` if any error arises the `start` lookup would be unregistered no matter what happens. I think a try-finally block is not necessary here.